File: code/stable/projectile/src/cpp/Calculations.cpp
#include "Calculations.hpp"

#include <algorithm>
#include <iostream>
#include <fstream>
#include <iterator>
#include <string>
#include <math.h>
#include <sstream>
#include <limits>
#include <vector>

using std::string;
using std::vector;
using std::ifstream;
using std::ofstream;

#include "InputParameters.hpp"

double func_t_flight(InputParameters &inParams) {
    return ((2 * (inParams.v_launch * sin(inParams.angle))) / 9.8);
}

double func_p_land(InputParameters &inParams) {
    return ((2 * (pow(inParams.v_launch, 2) * (sin(inParams.angle) * cos(inParams.angle)))) / 9.8);
}

double func_d_offset(InputParameters &inParams, double p_land) {
    return (p_land - inParams.p_target);
}

string func_s(InputParameters &inParams, double d_offset) {
    if ((fabs((d_offset / inParams.p_target)) < 2.0e-2)) {
        return "The target was hit.";
    }
    else if ((d_offset < 0)) {
        return "The projectile fell short.";
    }
    else if ((d_offset > 0)) {
        return "The projectile went long.";
    }
    else {
        throw("Undefined case encountered in function func_s");
    }
}

File: code/stable/projectile/src/csharp/Calculations.cs
/** \file Calculations.cs
    \author Samuel J. Crawford, Brooks MacLachlan, and W. Spencer Smith
    \brief Provides functions for calculating the outputs
*/
using System;

public class Calculations {
    
    /** \brief Calculates flight duration: the time when the projectile lands (s)
        \param inParams structure holding the input values
        \param g_vect gravitational acceleration (m/s^2)
        \return flight duration: the time when the projectile lands (s)
    */
    public static double func_t_flight(InputParameters inParams, double g_vect) {
        return 2 * inParams.v_launch * Math.Sin(inParams.theta) / g_vect;
    }
    
    /** \brief Calculates landing position: the distance from the launcher to the final position of the projectile (m)
        \param inParams structure holding the input values
        \param g_vect gravitational acceleration (m/s^2)
        \return landing position: the distance from the launcher to the final position of the projectile (m)
    */
    public static double func_p_land(InputParameters inParams, double g_vect) {
        return 2 * Math.Pow(inParams.v_launch, 2) * Math.Sin(inParams.theta) * Math.Cos(inParams.theta) / g_vect;
    }
    
    /** \brief Calculates distance between the target position and the landing position: the offset between the target position and the landing position (m)
        \param inParams structure holding the input values
        \param p_land landing position: the distance from the launcher to the final position of the projectile (m)
        \return distance between the target position and the landing position: the offset between the target position and the landing position (m)
    */
    public static double func_d_offset(InputParameters inParams, double p_land) {
        return p_land - inParams.p_target;
    }
    
    /** \brief Calculates output message as a string
        \param inParams structure holding the input values
        \param epsilon hit tolerance
        \param d_offset distance between the target position and the landing position: the offset between the target position and the landing position (m)
        \return output message as a string
    */
    public static string func_s(InputParameters inParams, double epsilon, double d_offset) {
        if (Math.Abs(d_offset / inParams.p_target) < epsilon) {
            return "The target was hit.";
        }
        else if (d_offset < 0) {
            return "The projectile fell short.";
        }
        else {
            return "The projectile went long.";
        }
    }
}
